// @angular/compiler-cli/src/ngtsc/path/src/types.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/path/src/types", ["require", "exports", "@angular/compiler-cli/src/ngtsc/path/src/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var util_1 = require("@angular/compiler-cli/src/ngtsc/path/src/util");
    /**
     * Contains utility functions for creating and manipulating `AbsoluteFsPath`s.
     */
    exports.AbsoluteFsPath = {
        /**
         * Convert the path `str` to an `AbsoluteFsPath`, throwing an error if it's not an absolute path.
         */
        from: function (str) {
            var normalized = util_1.normalizeSeparators(str);
            if (!util_1.isAbsolutePath(normalized)) {
                throw new Error("Internal Error: AbsoluteFsPath.from(" + str + "): path is not absolute");
            }
            return normalized;
        },
        /**
         * Assume that the path `str` is an `AbsoluteFsPath` in the correct format already.
         */
        fromUnchecked: function (str) { return str; },
        /**
         * Extract an `AbsoluteFsPath` from a `ts.SourceFile`.
         *
         * This is cheaper than calling `AbsoluteFsPath.from(sf.fileName)`, as source files already have
         * their file path in absolute POSIX format.
         */
        fromSourceFile: function (sf) {
            // ts.SourceFile paths are always absolute.
            return sf.fileName;
        },
    };
    /**
     * Contains utility functions for creating and manipulating `PathSegment`s.
     */
    exports.PathSegment = {
        /**
         * Convert the path `str` to a `PathSegment`, throwing an error if it's not a relative path.
         */
        fromFsPath: function (str) {
            var normalized = util_1.normalizeSeparators(str);
            if (util_1.isAbsolutePath(normalized)) {
                throw new Error("Internal Error: PathSegment.fromFsPath(" + str + "): path is not relative");
            }
            return normalized;
        },
        /**
         * Convert the path `str` to a `PathSegment`, while assuming that `str` is already normalized.
         */
        fromUnchecked: function (str) { return str; },
    };
});
