// @angular/compiler-cli/src/ngtsc/incremental/src/state.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/incremental/src/state", ["require", "exports", "tslib"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    /**
     * Accumulates state between compilations.
     */
    var IncrementalState = /** @class */ (function () {
        function IncrementalState(unchangedFiles, metadata) {
            this.unchangedFiles = unchangedFiles;
            this.metadata = metadata;
        }
        IncrementalState.reconcile = function (previousState, oldProgram, newProgram) {
            var e_1, _a, e_2, _b;
            var unchangedFiles = new Set();
            var metadata = new Map();
            // Compute the set of files that's unchanged.
            var oldFiles = new Set();
            try {
                for (var _c = tslib_1.__values(oldProgram.getSourceFiles()), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var oldFile = _d.value;
                    if (!oldFile.isDeclarationFile) {
                        oldFiles.add(oldFile);
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                }
                finally { if (e_1) throw e_1.error; }
            }
            try {
                // Look for files in the new program which haven't changed.
                for (var _e = tslib_1.__values(newProgram.getSourceFiles()), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var newFile = _f.value;
                    if (oldFiles.has(newFile)) {
                        unchangedFiles.add(newFile);
                        // Copy over metadata for the unchanged file if available.
                        if (previousState.metadata.has(newFile)) {
                            metadata.set(newFile, previousState.metadata.get(newFile));
                        }
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                }
                finally { if (e_2) throw e_2.error; }
            }
            return new IncrementalState(unchangedFiles, metadata);
        };
        IncrementalState.fresh = function () {
            return new IncrementalState(new Set(), new Map());
        };
        IncrementalState.prototype.safeToSkipEmit = function (sf) {
            if (!this.unchangedFiles.has(sf)) {
                // The file has changed since the last run, and must be re-emitted.
                return false;
            }
            // The file hasn't changed since the last emit. Whether or not it's safe to emit depends on
            // what metadata was gathered about the file.
            if (!this.metadata.has(sf)) {
                // The file has no metadata from the previous or current compilations, so it must be emitted.
                return false;
            }
            var meta = this.metadata.get(sf);
            // Check if this file was explicitly marked as safe. This would only be done if every
            // `DecoratorHandler` agreed that the file didn't depend on any other file's contents.
            if (meta.safeToSkipEmitIfUnchanged) {
                return true;
            }
            // The file wasn't explicitly marked as safe to skip emitting, so require an emit.
            return false;
        };
        IncrementalState.prototype.markFileAsSafeToSkipEmitIfUnchanged = function (sf) {
            this.metadata.set(sf, {
                safeToSkipEmitIfUnchanged: true,
            });
        };
        return IncrementalState;
    }());
    exports.IncrementalState = IncrementalState;
});
