// @angular/compiler-cli/ngcc/src/host/decorated_class.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/ngcc/src/host/decorated_class", ["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /**
     * A simple container that holds the details of a decorated class that has been
     * found in a `DecoratedFile`.
     */
    var DecoratedClass = /** @class */ (function () {
        /**
         * Initialize a `DecoratedClass` that was found in a `DecoratedFile`.
         * @param name The name of the class that has been found. This is mostly used
         * for informational purposes.
         * @param declaration The TypeScript AST node where this class is declared. In ES5 code, where a
         * class can be represented by both a variable declaration and a function declaration (inside an
         * IIFE), `declaration` will always refer to the outer variable declaration, which represents the
         * class to the rest of the program.
         * @param decorators The collection of decorators that have been found on this class.
         */
        function DecoratedClass(name, declaration, decorators) {
            this.name = name;
            this.declaration = declaration;
            this.decorators = decorators;
        }
        return DecoratedClass;
    }());
    exports.DecoratedClass = DecoratedClass;
});
