// @angular/core/schematics/migrations/injectable-pipe/angular/injectable_pipe_visitor.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/core/schematics/migrations/injectable-pipe/angular/injectable_pipe_visitor", ["require", "exports", "typescript", "@angular/core/schematics/utils/ng_decorators", "@angular/core/schematics/migrations/injectable-pipe/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    const ts = require("typescript");
    const ng_decorators_1 = require("@angular/core/schematics/utils/ng_decorators");
    const util_1 = require("@angular/core/schematics/migrations/injectable-pipe/util");
    /**
     * Goes through all of the descendant nodes of a given node and lists out all of the pipes
     * that don't have `@Injectable`, as well as their `@Pipe` decorator and the import declaration
     * from which we'd need to import the `Injectable` decorator.
     */
    class InjectablePipeVisitor {
        constructor(_typeChecker) {
            this._typeChecker = _typeChecker;
            /**
             * Keeps track of all the classes that have a `Pipe` decorator, but not `Injectable`, as well
             * as a reference to the `Pipe` decorator itself and import declarations from which we'll have
             * to import the `Injectable` decorator.
             */
            this.missingInjectablePipes = [];
        }
        visitNode(node) {
            switch (node.kind) {
                case ts.SyntaxKind.ClassDeclaration:
                    this._visitClassDeclaration(node);
                    break;
            }
            ts.forEachChild(node, node => this.visitNode(node));
        }
        _visitClassDeclaration(node) {
            if (!node.decorators || !node.decorators.length) {
                return;
            }
            const ngDecorators = ng_decorators_1.getAngularDecorators(this._typeChecker, node.decorators);
            const pipeDecorator = ngDecorators.find(decorator => decorator.name === 'Pipe');
            const hasInjectableDecorator = !ngDecorators.some(decorator => decorator.name === util_1.INJECTABLE_DECORATOR_NAME);
            // Skip non-pipe classes and pipes that are already marked as injectable.
            if (pipeDecorator && hasInjectableDecorator) {
                const importNode = pipeDecorator.importNode;
                const namedImports = importNode.importClause && importNode.importClause.namedBindings;
                const needsImport = namedImports && ts.isNamedImports(namedImports) &&
                    !namedImports.elements.some(element => element.name.text === util_1.INJECTABLE_DECORATOR_NAME);
                this.missingInjectablePipes.push({
                    classDeclaration: node,
                    importDeclarationMissingImport: needsImport ? importNode : null,
                    pipeDecorator: pipeDecorator.node
                });
            }
        }
    }
    exports.InjectablePipeVisitor = InjectablePipeVisitor;
});
